### PR TITLE
RSA4c+d

### PR DESCRIPTION
**Pull request by withakay**
Changes and tests to cover RSA4c (1,2,3) and RSA4d

### CODE REVIEW

File: src/IO.Ably.Tests.Shared/AuthTests/AuthSandboxSpecs.cs [hunk 1]
```diff
@@ -261,66 +264,143 @@ private static TokenParams CreateTokenParams(Capability capability, TimeSpan? tt
...
+                }
+
+                TaskCompletionAwaiter tca = new TaskCompletionAwaiter(1000);
+                var realtimeClient = await GetRealtimeClient(protocol, Options);
+
+                realtimeClient.Connect();
+                await realtimeClient.WaitForState(ConnectionState.Connected);
+                realtimeClient.Connection.On(change =>
+                {
+                    // this callback should not be called
```
Comment (funkyboy) [inline src/IO.Ably.Tests.Shared/AuthTests/AuthSandboxSpecs.cs]: If I am reading this correctly, there should be a `fail` statement here. No?

Comment (withakay) [reply src/IO.Ably.Tests.Shared/AuthTests/AuthSandboxSpecs.cs]: @cesare the failure is on line 353 `(await tca.Task).Should().BeFalse(context);`

File: src/IO.Ably.Shared/AblyAuth.cs [hunk 2]
```diff
@@ -213,9 +213,17 @@ private void SetAuthMethod()
                 }
                 catch (Exception ex) when (shouldCatch)
                 {
+                    HttpStatusCode? statusCode = null;
+                    int code = 80019;
+                    if (ex is AblyException ablyException)
```
Comment (paddybyers) [inline src/IO.Ably.Shared/AblyAuth.cs]: I don't think this implements the requirement. The outer error has code 80019 and it wraps the original error as `cause`. I don't know why this references 40300.

Comment (paddybyers) [reply src/IO.Ably.Shared/AblyAuth.cs]: Where are you getting that text from?

The requirement I'm looking at is in https://github.com/ably/docs/blob/integration-spec-1-1/content/client-lib-development-guide/features.textile

The required check for RSA4d is `statusCode == 403`, not `code == xx`. And either way once you have the inner error you generate the ErrorInfo to return with `code = 80019, statusCode = <inner err>.statusCode, cause = <inner err>`. The difference when RSA4d applies is in the Realtime case you transition to `failed`.

Comment (withakay) [reply src/IO.Ably.Shared/AblyAuth.cs]: That text is from the from the Google Sheet titled "Ably .NET Realtime Client Library Spec - v1.1"

Comment (withakay) [reply src/IO.Ably.Shared/AblyAuth.cs]: @paddybyers fixed by https://github.com/ably/ably-dotnet/pull/212/commits/1d1d88a602c12d960807272fb6547ba5e4378ceb & https://github.com/ably/ably-dotnet/pull/212/commits/7572963e1334edaeec73c858730791962cc9cee9

File: src/IO.Ably.Tests.Shared/Realtime/ConnectionSandBoxSpecs.cs [hunk 3]
```diff
@@ -1001,55 +997,10 @@ public async Task WhenFakeAuthMessageReceived_ShouldAttemptTokenRenewal(Protocol
...
-            var initialClientId = client.ClientId;
-
-            client.Connection.Once(ConnectionEvent.Disconnected, state2 =>
-            {
-                client.Connection.Once(ConnectionEvent.Connected, state3 =>
-                {
-                    reconnectAwaiter.SetCompleted();
-                });
-            });
+            await Task.Delay(1000);
```
Comment (paddybyers) [inline src/IO.Ably.Tests.Shared/Realtime/ConnectionSandBoxSpecs.cs]: Is this a "magic" delay?

Comment (withakay) [reply src/IO.Ably.Tests.Shared/Realtime/ConnectionSandBoxSpecs.cs]: There are points where some internal events are raised and if I recall it was possible to hit the next line before they had been handled, so this delay allow for that. So yes, somewhat magical.

**Pull request merged**